1130 combobox

davidicus opened the pull request:
Closes #1264 closes #1265 

**Summary**

Extend the Carbon Combobox component to implement dynamic input and multi-value input

**Change List (commits, features, bugs, etc)**

- Duplicated Carbon's combobox component in CarbonCombobox.jsx. This will be removed once we have migrated to `carbon-components-react@7.13.0`
- Created our own ComboBox Component in ComboBox.jsx
- Created test for both ComboBox.js and CarbonComboBox.jsx
- added styles to _combo-box.scss

**Acceptance Test (how to verify the PR)**

- Go to the experimental Combobox story and play with each one. The existing stories should work the same as base carbon with the addition of adding the new item to the dropdown list
- Go to multi-value story and make sure tags are added and removed  properly.
- Actions pannel should show the `onChange` callback being called with the appropriate value

davidicus commented on src/utils/__tests__/__snapshots__/publicAPI.test.js.snap:
```diff
@@ -13273,159 +13273,22 @@ Map {
       },
     },
   },
-  "Accordion" => Object {
```
Not quite sure what these changes are. Will have to check what's happening here.

davidicus replied:
OK, these were not removed just moved down the file. It is good to go.

bjornalm commented:
Looks like we are missing some padding in RTL
![image](https://user-images.githubusercontent.com/5167091/84917359-8253e600-b0bf-11ea-8385-1474fbeb15f8.png)

bjornalm commented:
Keyboard navigation is not working for selections in the multi-value tags Dropdown

bjornalm commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+      // Check if there is already a tag for this value and return it
+      const filteredTags = tagItems.filter(x => itemToString(x) === currentValue);
+      // Check if value is part of items array
+      const filteredItems = listItems.filter(x => itemToString(x) === currentValue);
```
same here, there is a lot of `filteredItems.length < 1` and `filteredItems[0]`, we could use a constant like `matchedItem` that either is undefined or is the item

bjornalm commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+      // create new item to add items array and tags array
+      const newItem = {
+        id: `id-${uid}`,
+        text: currentValue || '',
```
What kind of trimmed currentValue is expected to be falsy here?

davidicus replied:
This was a holdover. I can remove.

davidicus commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+  useEffect(() => {
+    // If there are tags then clear and focus the input
+    if (tagItems.length >= 1) {
+      setSelectedItem('');
```
we only clear for the multivalue feature. I guess I can change to `hasMultiValue`

bjornalm replied:
My question was more about ` setSelectedItem('');` vs ` setSelectedItem(null);` and if we really need to use both?

davidicus replied:
Do not need both. Changed to null

davidicus commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+    // Only apply logic if user hit enter
+    if (evt.key === 'Enter' && currentValue !== '' && currentValue !== undefined) {
+      // Check if there is already a tag for this value and return it
+      const filteredTags = tagItems.filter(x => itemToString(x) === currentValue);
```
I am hard pressed to believe that. double bang is easier to read and decipher.  if the use is ambiguous we can do something like `.length > 0` or some such

davidicus commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+  // Array that populates tags
+  const [tagItems, setTagItems] = useState([]);
+
+  useEffect(() => {}, []);
```
none at all.

davidicus commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,204 @@
+              key={`tag-${item?.id}-${idx}`}
+              filter
+              onClose={e => handleOnClose(e)}
+              title={closeButtonText}
```
its used for the aria-label by carbon

davidicus commented on src/components/ComboBox/ComboBox.test.jsx:
```diff
@@ -0,0 +1,235 @@
+  });
+
+  afterEach(() => {
+    cleanup();
```
No, removed

tay1orjones commented:
> Duplicated Carbon's combobox component in CarbonCombobox.jsx. This will be removed once we have migrated to carbon-components-react@7.13.0

I have created issue https://github.com/IBM/carbon-addons-iot-react/issues/1323 to track the update to Carbon v10.13.0 and included this there as an action item.

bjornalm commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -71,52 +69,52 @@ const ComboBox = ({
     if (evt.key === 'Enter' && currentValue !== '' && currentValue !== undefined) {
       // Check if there is already a tag for this value and return it
-      const filteredTags = tagItems.filter(x => itemToString(x) === currentValue);
+      const hasNoFilteredTag = tagItems.filter(x => itemToString(x) === currentValue).length < 1;
```
I think the comment above here, and possibly some other comments can be removed now or at least modified

davidicus commented:
> > Duplicated Carbon's combobox component in CarbonCombobox.jsx. This will be removed once we have migrated to carbon-components-react@7.13.0
> 
> I have created issue #1323 to track the update to Carbon v10.13.0 and included this there as an action item.

I actually have had to make some additional changes to the Carbon component. I will opening up a PR soon and I will link to it in this issue.

davidicus commented:
Added extra test to 

> Keyboard navigation is not working for selections in the multi-value tags Dropdown

Added test to account for keyboard navigation

bjornalm commented:
I did notice a new issue when testing the keyboard selection. It adds both the text value and the matching item when I hit enter.

![image](https://user-images.githubusercontent.com/5167091/84993064-0aca9900-b149-11ea-9041-f6797335987f.png)

davidicus commented:
> I did notice a new issue when testing the keyboard selection. It adds both the text value and the matching item when I hit enter.
> 
> ![image](https://user-images.githubusercontent.com/5167091/84993064-0aca9900-b149-11ea-9041-f6797335987f.png)

Good catch @bjornalm. Speaking with @lukefirth about the UX issue this surfaces.

davidicus commented:
Hey @bjornalm I had to modify pretty significantly. May want to take another look.

sstone2423 commented on src/components/ComboBox/CarbonComboBox.jsx:
```diff
@@ -0,0 +1,582 @@
+import { Checkmark16, WarningFilled16 } from '@carbon/icons-react';
+import ListBox from 'carbon-components-react/es/components/ListBox';
+
+const keys = {
```
I believe theres some keyCode constants already in the utils. Could you move this over there?

davidicus replied:
This is from Carbon. With plans on pushing these changes back. There are some tweaks to the original Carbon component which is why we are temporarily carrying this component. I would rather not abstract these to another file only to have to delete them when this file is deleted.

sstone2423 commented on src/components/ComboBox/CarbonComboBox.jsx:
```diff
@@ -0,0 +1,582 @@
+import React from 'react';
```
is this the original ComboBox that you're leaving unaltered?

davidicus replied:
This is an altered version of the Carbon component.

sstone2423 commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,218 @@
+
+  const handleOnClose = e => {
+    // Get close target's text
+    const closedValue = e.currentTarget.parentNode.children[0].textContent;
```
Does there need to be any null checks here? I'm not sure how likely it would be that the textContent doesn't exist?

davidicus replied:
Unlikely that most of these will be nulls but will put some on the children. in case of a markup change.

sstone2423 commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,218 @@
+    });
+    // Send new value to users onChange callback
+    onChange([...tagItems]);
+    e.currentTarget.parentNode.parentNode.parentNode.firstChild.children[0].children[1].focus();
```
Same null check comment as above

sstone2423 commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,218 @@
+    // If there are tags then clear and focus the input
+    if (hasMultiValue) {
+      setSelectedItem(null);
+      // comboRef.current.textInput.current.focus();
```
is this needed anymore?

davidicus replied:
No. will remove

sstone2423 commented on src/components/ComboBox/ComboBox.jsx:
```diff
@@ -0,0 +1,218 @@
+  onChange,
+  items,
+  downshiftProps,
+  ...comboProps
```
could these props be listed?

tay1orjones commented:
Overrriding status checks - they're passing. Intermittent API issues look like the culprit.

tay1orjones commented:
:tada: This PR is included in version 2.99.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react)
- [GitHub release](https://github.com/IBM/carbon-addons-iot-react/releases/tag/v2.99.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
